### DISCUSSION TITLE
Better handling of optional dependencies

### DISCUSSION THREAD

**Discussion by VorpalBlade** [CyberShadow/aconfmgr#152] [Ideas]
Quoting the readme:

> * Packages that are only optional dependencies of other packages must be listed explicitly, otherwise they will be pruned.
> * `aconfmgr apply` removes unlisted packages by unpinning them (setting their install reason as "installed as a dependency"), after which it prunes all orphan packages. If the package is still required by another package, it will remain on the system (until it is no longer required); otherwise, it is removed.

There are two issues with this.

1. An initial `save` followed by `apply` is not idempotent if there are any packages in this state.
2. I find it useful to track when a dependency has been installed as a dependency of a package, even if it is only an optional dependency. If I uninstall the package that depends on this optional package `pacman -Qtd` will tell me about the no longer needed dependency. This only works if the package is not marked as explicit however. Without this, there is a high risk I will forget about the optional dependency and leave it installed.

I propose a new `AddPackageAsDependency` (or similar, I don't care about the name of it), or perhaps a flag to `AddPackage` that will allow marking the package as installed as a dependency to satisfy both `pacman` and `aconfmgr`

**Comment by CyberShadow** [comment]
> An initial `save` followed by `apply` is not idempotent if there are any packages in this state.

Yeah. It's packages which are installed "as dependencies" (`--asdeps`), but there's nothing actually requiring them. The problem from aconfmgr's side is that it's difficult to calculate the exact set of these packages, taking into account transitivity and things like `provides`.

I think as a compromise we could make aconfmgr print a warning after `save` if it detects that there exists at least one package in this state, to mention that the configuration is not yet fully representative of the system state.

> I find it useful to track when a dependency has been installed as a dependency of a package, even if it is only an optional dependency. If I uninstall the package that depends on this optional package `pacman -Qtd` will tell me about the no longer needed dependency. This only works if the package is not marked as explicit however. Without this, there is a high risk I will forget about the optional dependency and leave it installed.

I think this particular problem has a much simpler practical solution: list dependencies of a package right where you have the package itself. Most non-trivial parts of the configuration are going to have additional declarations for things like configuration or ignore files, and dependencies are a part of that. There's no harm in having multiple `AddPackage` lines for the same package, so dependencies can be repeated for different dependent packages (though it does make it easy to accidentally omit a dependency which should be specified).

See also:

- #58
- #76
